x-pack/plugins/security_solution/server/lib/detection_engine/signals/alert_suppression/group_and_bulk_create.ts

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type moment from 'moment';
+
+import type * as estypes from '@elastic/elasticsearch/lib/api/types';
+
+import { withSecuritySpan } from '../../../../utils/with_security_span';
+import { buildTimeRangeFilter } from '../build_events_query';
+import type {
+  EventGroupingMultiBucketAggregationResult,
+  GroupAndBulkCreateParams,
+  GroupAndBulkCreateReturnType,
+} from '../types';
+import { addToSearchAfterReturn, getUnprocessedExceptionsWarnings } from '../utils';
+import type { ThrottleBuckets } from '../../rule_types/factories/utils/wrap_throttled_alerts';
+import { wrapThrottledAlerts } from '../../rule_types/factories/utils/wrap_throttled_alerts';
+import { buildGroupByFieldAggregation } from './build_group_by_field_aggregation';
+import { singleSearchAfter } from '../single_search_after';
+
+export interface BucketHistory {
+  key: Record<string, string | number>;
+  endDate: Date;
+}
+
+/**
+ * Builds a filter that excludes documents from existing buckets.
+ */
+export const buildBucketHistoryFilter = ({
+  bucketHistory,
+  primaryTimestamp,
+  secondaryTimestamp,
+  from,
+}: {
+  bucketHistory: BucketHistory[];
+  primaryTimestamp: string;
+  secondaryTimestamp: string | undefined;
+  from: moment.Moment;
+}): estypes.QueryDslQueryContainer[] | undefined => {
+  if (bucketHistory.length === 0) {
+    return undefined;
+  }
+  return [
+    {
+      bool: {
+        must_not: bucketHistory.map((bucket) => ({
+          bool: {
+            filter: [
+              ...Object.entries(bucket.key).map(([field, value]) => ({
+                term: {
+                  [field]: value,
+                },
+              })),
+              buildTimeRangeFilter({
+                to: bucket.endDate.toISOString(),
+                from: from.toISOString(),
+                primaryTimestamp,
+                secondaryTimestamp,
+              }),
+            ],
+          },
+        })),
+      },
+    },
+  ];
+};
+
+export const filterBucketHistory = ({
+  bucketHistory,
+  fromDate,
+}: {
+  bucketHistory: BucketHistory[];
+  fromDate: Date;
+}) => {
+  return bucketHistory.filter((bucket) => bucket.endDate > fromDate);
```

Comment:
it looks like `endDate` in bucketHistory is an ISO string, but fromDate is of `Date` type. So, this comparison always will be false

Reply:
`endDate` is defined as `Date` [here](https://github.com/elastic/kibana/pull/142686/files#diff-c4f72d9aebfdcf91b8502f6d1bf82695ef63d6bd44337517a731e3749ebcb648R27) with a unit test to verify that it filters out 1 test bucket from a set of 2

Reply:
In unit test, endDate is of `Date` type, so it works as expected.

While debugging rule execution, I've noticed actual type of endDate that coming in bucketHistory is a string.
Here is bucketHistory as it getting processed for filtering in `filterBucketHistory`
```JSON
[
  {
    "key": {
      "host.name": "host-11",
      "labels.groupId": "group-0"
    },
    "endDate": "2022-11-07T09:53:53.217Z"
  },
  {
    "key": {
      "host.name": "host-11",
      "labels.groupId": "group-1"
    },
    "endDate": "2022-11-04T09:50:53.217Z"
  },
  {
    "key": {
      "host.name": "host-11",
      "labels.groupId": "group-2"
    },
    "endDate": "2022-11-04T09:50:53.217Z"
  },
  {
    "key": {
      "host.name": "host-11",
      "labels.groupId": "group-3"
    },
    "endDate": "2022-11-07T09:53:53.217Z"
  }
]
```

Seems like, it's just parsed JSON from task instance state, which is stored as string in ES

```JSON
{
  "_index": ".kibana_task_manager_8.6.0_001",
  "_id": "task:893d8550-5f56-11ed-8c39-ab0585437160",
  "_version": 17,
  "_seq_no": 44389,
  "_primary_term": 1,
  "found": true,
  "_source": {
    "migrationVersion": {
      "task": "8.5.0"
    },
    "task": {
      "retryAt": null,
      "runAt": "2022-11-08T11:31:35.723Z",
      "startedAt": null,
      "params": """{"alertId":"893d8550-5f56-11ed-8c39-ab0585437160","spaceId":"default","consumer":"siem"}""",
      "ownerId": null,
      "enabled": true,
      "schedule": {
        "interval": "3m"
      },
      "taskType": "alerting:siem.queryRule",
      "scope": [
        "alerting"
      ],
      "traceparent": "",
      "state": """{"alertTypeState":{"throttleGroupHistory":[{"key":{"host.name":"host-11","labels.groupId":"group-0"},"endDate":"2022-11-07T09:53:53.217Z"},{"key":{"host.name":"host-11","labels.groupId":"group-1"},"endDate":"2022-11-04T09:50:53.217Z"},{"key":{"host.name":"host-11","labels.groupId":"group-2"},"endDate":"2022-11-04T09:50:53.217Z"},{"key":{"host.name":"host-11","labels.groupId":"group-3"},"endDate":"2022-11-07T09:53:53.217Z"}]},"alertInstances":{},"previousStartedAt":"2022-11-08T11:28:35.723Z"}""",
      "scheduledAt": "2022-11-08T11:28:34.465Z",
      "attempts": 0,
      "status": "idle"
    },
    "references": [],
    "updated_at": "2022-11-08T11:28:37.875Z",
    "coreMigrationVersion": "8.6.0",
    "created_at": "2022-11-08T11:16:17.165Z",
    "type": "task"
  }
}
```

Reply:
Ah I see, that's a great catch. I'll update the rule `state` type so the `endDate` is correctly represented as `string` and turn it into `Date` when needed.